### *“Learning to write programs stretches your mind, and helps you think better, [it] creates a way of thinking about things that I think is helpful in all domains.”* 

*–Bill Gates, Co-Founder, Microsoft*

I agree! It does not matter what your major is - Programming builds analytical skills tansferrable to any field."
    
---

## <span style="color:red">Your Name: Daniel DeLuca</red>

# Lab B: The Basics of Python - Types and Variables!

---

This lab was adapted from an IBM course ("Python for Data Science") on the Cognitive Class website. Thank you to the original author <a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a> who is a Data Scientist at IBM for creating the initial lab. 

The goal of this lab is to get you familiar with the basics of Python programming.  Read the lab very carefully, taking care to perform each task as indicated in the code blocks.  There is some overlap with Lab A, but I'm just trying to ease you in.


## NOTE: If You Get Overhelmed  

You will get better with Python as you go through the course.  Just ask a lot of questions from me and your lab partner. If you start this lab and get overwhelmed, put on your headphones, watch these videos, and then retry the lab.  

* [Cognitive Class Video on Expressions in Python](https://youtu.be/K7-oodBQM2o)

* [Cognitive Class Video on Strings](https://youtu.be/D9yUVAmjLuc)

## NOTE: Don't Abuse the Answers

The answers to the exercises are embedded in this lab.  It would be a TERRIBLE idea to just look at the answer right away.  Try the question several times first.  If you're in the lab ask questions of your neighbors and me.  Look at the answer only when you think you have it right and want to check yourself.

---

## 1. Say "Hello" to the world in Python

The first thing we always learn when programming is how to print out information to the screen.  This one line of code will ensure that we know how to print a string in output and how to execute code within cells in a notebook.

---

<div class="alert alert-success alertsuccess" style="margin-top: 20px">
[Tip]: To execute the Python code in the code cell below, use the run button aboce or click on the cell to select it and press <kbd>Shift</kbd> + <kbd>Enter</kbd>.  
   <br/>
       <br/>
(FYI: <kbd>Control</kbd> + <kbd>Enter</kbd> also works!)
</div>

---

In [1]:
# Try your first Python output (Shift+Enter executes the cell)

print('Hello, Students!')

Hello, Students!


<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [Tip:] <code>print()</code> is a function. You passed the string <code>'Hello, Python!'</code> as an argument to instruct Python on what to print.
</div>
<hr/>

<h2 id="comments">2. Writing comments in Python</h2>

We often want to write comments in our code cells, just to explain what the code is doing.

To write comments in Python, use the hashtag symbol ``#`` before writing your comment. When you run your code, Python will ignore everything past the ``#`` on a given line.


In [2]:
# Run this cell (Remember: Shift+Enter executes the cell)

print('Hello, Python!') # This line prints a string
# print('Hi')  -  Do you see this printed out? 

Hello, Python!


<hr/>
<h2 id="errors">3. Errors in Python</h2>

<p>Python let's you know when you make certain types of mistake.  You should read the error message you are given to determine WHERE the mistake is and how you may correct it.</p>

In [6]:
# Run this cell to see the error. (Shift+Enter executes the cell.)

#Print is not spelled correctly: uh-oh! 
print("Hello, Python!")

Hello, Python!


Syntax is incredibly important in Python and any coding language.  For example, whenever you want to print TEXT from a code cell, the TEXT must START AND END in with a quote.  Note that the computer science word for TEXT is **STRING**.

In [5]:
# Run this cell to see the error. (Remember: Shift+Enter executes the cell)

print("Hello, Python!")

Hello, Python!


To help you, the notebook shows all strings in red.    

In the code block above, the coder forgot the second quote, so the notebook thinks the second parenthesis is considered part of the text and is thus is colored RED.  

If you add the ending quote, the second parenthesis will turn back to black.  This means that the parenthesis is now recognized as code to end the print function.

### Watch this video to listen to me explain the red error output above.

[Video Explaining these errors - Click here.](https://desales.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=505351e2-52bf-4c35-89df-acc600e3534b)

---
## **Question 3A!!!!!!!!!!!!!!!!!!!** 

Fix the errors in the above 2 code cells so that they run without errors.

---

### Does Python know about your error before it runs your code?

Python is what is called an <em>interpreted language</em> which means Python interprets your script line by line as it executes it.  So Python will stop executing the entire program once it encounters an error.  

As a comparison, there are languages called <em> compiled languages</em> which examine your entire program at compile time, and are able to warn you about a whole class of errors prior to execution.

In [7]:
# Print string and error to see the running order.  Notice which
#statements are printed and which are not.

print("This will be printed")
pint("This will cause an error")  #Error here: pint instead of print - This line will therefore not print.
print("This will NOT be printed") #Do you see this printed out?

This will be printed


NameError: name 'pint' is not defined

You should have noticed that only the first line is printed out above.  Because the second has an error, neither the second nor third line completes its print.

---
## **Question 3B!!!!!!!!!!!!!!!!!!!** 

<p>In the code cell below, use the <code>print()</code> function 3 times to print out your name, major, and hometown.  For example, Joe Biden would print out the following.</p>
<code>Joe Biden
History and Political Science
Scranton, PA</code>

In [ ]:
# Write your code below and press Shift+Enter to execute 

In [9]:
print("DANIEL DELUCA")
print("SUPPLY CHAIN MANAGEMENT")
print("BETHLEHEM, PA")


DANIEL DELUCA
SUPPLY CHAIN MANAGEMENT
BETHLEHEM, PA


Double-click __here__ for the solution.

<!-- Joe Biden would have typed the following code:

print("Joe Biden")
print("History and Political Science")
print("Scranton, PA")

-->
<hr/>

<h2 id="python_error">4. New Line Characters:  Using one print statement to write on multiple lines</h2>

You may think you have to write 3 print statements to print the statement below.  

<code>
A
B
C
</code>

<br/>

However, we can instead use \\n within a STRING to print a line break.  Using \\n within a string means \"please insert a newline here\".

In [10]:
#The following prints the characters A, B, and C on its own line.  Try it! (Remember: Shift+Enter executes the cell)

print("A\nB\nC")

A
B
C


---
## **Question 4!!!!!!!!!!!!!!!!!!!** 
    

<p>Print the following using ONE print statement with NEWLINE characters.</p>
print("Joe Biden\nHistory and Political Science/nScranton, PA")

In [13]:
print("Joe Biden\nHistory and Political Science \nScranton, PA")#Write your code here and run the cell to test it.  (Remember: Shift+Enter executes the cell)

Joe Biden
History and Political Science 
Scranton, PA


Double-click __here__ for the solution.

<!-- Joe Biden would have typed the following code:

print("Joe Biden\nHistory and Political Science\nScranton, PA")

-->

<hr>

<h2 id="types_objects">5. Numbers</h2>

Quantitative information arises everywhere in data science. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.2500` evaluates to the number 3.25. (Run the cell and see.)

In [14]:
3.2500

3.25

In [15]:
#Run this cell and see what does and does not print
print(2)
3
4

2


4

Above, you should have noticed that:

* 4 is the value of the last expression and is thus printed
* 2 is printed because you specified this with the "print" command, 
* but 3 is lost forever because it did not appear in a ``print`` statement nor was it the last value in the cell.



Notice that we didn't have to use the `print` function to print anything to the screen.  When you run a notebook cell, if the last line has a value, then the Jupyter notebook automatically prints out that value for you. However, it won't print out prior lines automatically.

---
## **Question 5!!!!!!!!!!!!!!!!!!!** 
    
Run the code below and verify that only the 9 will print out. Update the code so that the number on each of the lines is printed.

In [18]:
print("7")
print("8")
9

7
8


9

Double-click __here__ for the solution.

<!-- 

print("7")
print("8")
9

OR YOU COULD DO:

print("7\n8")
9

OR

print("7\n8\n9")

-->
<hr/>

## 6. Names
In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we just say, "Bear!"

Similarly, an effective strategy for writing code is to define names for data as we compute it, like a lawyer would define terms for complex ideas at the start of a legal document.

In Python, we do this with *assignment statements*. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right.

In [21]:
#Run this cell
ten = 3 * 2 + 4

When you run that cell, Python first evaluates the first line.  It computes the value of the expression `3 * 2 + 4`, which is the number 10.  Then it gives that value the name `ten`.  At that point, the code in the cell is done running.

After you run that cell, the value 10 is bound to the name `ten`:

In [22]:
#Run this cell now and see what happens
ten

10

The statement `ten = 3 * 2 + 4` is not asserting that `ten` is already equal to `3 * 2 + 4`, as we might expect by analogy with math notation.  Rather, that line of code changes what `ten` means; it now refers to the value 10, whereas before it meant nothing at all.

If the designers of Python had been ruthlessly pedantic, they might have made us write:

    define the name ten to hereafter have the value of 3 * 2 + 4 

instead to accomplish the same thing.  You will probably appreciate the brevity of "`=`"!  But keep in mind that this is the real meaning.

---
## **Question 6A!!!!!!!!!!!!!!!!!!!** 
    
Try writing code in the cell below that uses a name (like `foo` or `eleven`) that hasn't been assigned to anything.  You'll see an error!

In [23]:
#Write your code here
eleven

NameError: name 'eleven' is not defined

<hr>

A common pattern in Jupyter notebooks is to assign a value to a name and then immediately evaluate the name in the last line in the cell so that the value is displayed as output. 

In [24]:
#Run this cell and see what happens!

close_to_pi = 355/113
close_to_pi

3.1415929203539825

FYI:  We often refer to close_to_pi as a variable.  It is a variable that stores the value ``3.1415929203539825...``.

Another common pattern is that a series of lines in a single cell will build up a complex computation in stages, naming the intermediate results.

In [25]:
#Run this cell and see what happens!

bimonthly_salary = 840

monthly_salary = 2 * bimonthly_salary   #monthly_salary is a variable name that holds an intermediate results

number_of_months_in_a_year = 12

yearly_salary = number_of_months_in_a_year * monthly_salary

yearly_salary        #yearly_salary is a variable name that holds the final result of interest

20160

Variable names in Python can have letters (upper- and lower-case letters are both okay and count as different letters), underscores, and numbers.  The first character can't be a number (otherwise a name might look like a number).

Other than those rules, what you name something doesn't matter *to Python*.  For example, this cell does the same thing as the above cell, except everything has a different name:

In [ ]:
a = 840
b = 2 * a
c = 12
d = c * b
d

**HOWEVER**, names are very important for making your code *readable* to yourself and others.  The cell above is shorter, but it's totally useless without an explanation of what it does.

According to a famous joke among computer scientists, naming things is one of the two hardest problems in computer science.  

(The other two are cache invalidation and "off-by-one" errors.  And people say computer scientists have an odd sense of humor...)

---

## **Question 6B!!!!!!!!!!!!!!!!!!!** 

Assign the name `seconds_in_a_decade` to the number of seconds between midnight January 1, 2010 and midnight January 1, 2020.

<span style="color:red">DO THIS WITHOUT USING ANY TYPE OF CALCULATOR/GOOGLE.</span> The point is for you to start using Jupyter Notebooks and Python as a calculator.  

In [26]:
# Change the next line so that it computes the number of
# seconds in a decade and assigns that number the name
# seconds_in_a_decade.
seconds_in_a_decade = 60 * 60 * 24 * 365 * 10

# We've put this line in this cell so that it will print
# the value you've given to seconds_in_a_decade when you
# run it.  You don't need to change this.
seconds_in_a_decade

315360000

<hr>

<h2 id="types_objects">7. Variables and Their Types</h2>

We used the word "names" above to describe english words that refer to values.  However, as mentioned beforehand, just like you did in math class, we use the term "variables" to refer to "names".

Variables can have many different types.  Often we are working with INTEGERS, FLOATS, or STRINGS.
<ul>
<li>Whole numbers are refered to as INTEGERS: 10, 15, 318.</li>
<li>Numbers with decimals are referred to as FLOATS: 3.1415926, 2.58, 101.3.</li>
<li>Another important type is STRING.  A string is just a word or a series of letters, like "HOWDY" or "Python" or "What?!!".</li>
</ul> 

**Directions:**  Execute the cells below (shift + enter) to see what is printed out.  

In [27]:
#Here we create an integer variable/name
x = 7
print(x)

 #The type function tells us the type of variable that x is.
print(type(x))

7
<class 'int'>


You should have seen ``<class 'int'>`` print out as part of the output.  That tells us that the variable ``x`` is seen as an integer.

In [28]:
#Here we create a float variable and print its type.
y = 3.14
print(y)
print(type(y))


3.14
<class 'float'>


You should have seen ``<class 'float'>`` print out as part of the output above.  That tells us that the variable ``y`` is seen as number with decimal values.

In [29]:
#Here we create a string variable and print its type
sentence = "DeSales MACS department rocks!"
print(sentence)
print(type(sentence))

DeSales MACS department rocks!
<class 'str'>


You should have seen ``<class 'str'>`` print out as part of the output above.  That tells us that the variable ``sentence`` is seen as a string or text.

## One note on strings:  

* In Python, it does not matter whether you use single quotes or double quotes to create strings.

In [30]:
language1 = "Spanish"
print(type(language1))   #language1 is a string

language2 = 'Spanish'
print(type(language2)) #Is language2 a string too?

<class 'str'>
<class 'str'>


In [31]:
#Do these print statementes cause anything different to happen?
print("Hello my friend!")
print('Hello my friend!')

Hello my friend!
Hello my friend!


So remember:  Whether you use single or double quotes to designate a string does not matter.


<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [Tip:] Note that strings can be represented with single quotes (<code>'cat'</code>) or double quotes (<code>"cat"</code>), but you can't mix both (e.g., <code>"cat'</code>).
</div>


<hr/>
<h2>8. Manipulating Variables</h2>

<p>We can manipulate variables by adding/subtracting/multiplying/dividing them.</p>  

**Directions:** Guess what is printed out below. Then execute the cells to check your guess. (Remember: shift+enter executes the cell)

In [32]:
#Here are the length of 3 sides of a triangle
triangle_side1 = 3;
triangle_side2 = 4;
triangle_side3 = 5;

#computer the perimeter - add the sides!
perimeter = triangle_side1 + triangle_side2 + triangle_side3 

#Print out the perimeter. Guess what is printed.
print(perimeter)

12


In [33]:
#To multiply, we use the "*" symbol

a = 3
b =  10
print(a * b)  #what is printed out?

30


In [34]:
#Here's another example of manipulating variables.
time_in_seconds = 30
time_in_minutes = time_in_seconds / 60  #Divide seconds by 360 to convert to minutes.

print(time_in_seconds)  #prints 30
print(time_in_minutes)  #what should this print?   Run this code to see!


30
0.5


You can change the value of variables within your code.  In the example below, x starts at the value of 5 and then we change the value of x.

In [35]:
#Here x is 5
x = 5
print(x)

#Change the value of x and reprint.
x = 7
print(x)

#What do you think the new value of x is with the code below?  Guess and run to find out!
x = x + 4
x 

5
7


11


We can also "add" strings! 

Doing so just combines the 2 strings into 1 longer string and this is called __STRING CONCATENATION__.

In [36]:
#Let's create 2 stings.
word1 = "DeSales"
word2 = "University"

#And now we concatenate them by adding them - What will be printed?  Run the cell to see!
bigger_word = word1 + word2
bigger_word

'DeSalesUniversity'

Notice that no space was added between word1 and word2 above.  Let's fix that.

In [37]:
#Now we concatenate the strings with a space in the middle. 
print(word1 + " " + word2

SyntaxError: unexpected EOF while parsing (Temp/ipykernel_23720/394938321.py, line 2)

So recapping:  String concatenation is the processing of just taking one string on to the end of another and is done by use a PLUS sign between strings.

---
 
## **Question 8A!!!!!!!!!!!!!!!!!!!** 


In the code block below, create a variable called yards which is equal to the value of the feet variable when it is converted to yards.  Then print the value of the yards variable.

In [41]:
feet = 10
yards = 3 / feet
print(yards)

#Create a variable called yards here

#Print out the value of the yards variable here

0.3


Double-click __here__ for the solution.

<!-- 
yards = feet / 3  #There are 3 feet in every yard
print(yards)
-->

---

## **Question 8B!!!!!!!!!!!!!!!!!!!** 

The code cell below has two variables.  Use string concatenation to print out 
"Philadelphia, PA" (with the comma and spaces just like this) to the screen

In [46]:
city = "Philadelphia"
state = "PA"
print (city + ", " + state)

Philadelphia, PA


Double-click __here__ for the solution.

<!-- 
new_phrase = city + ", " + state
print(new_phrase)
-->

 <hr/> <h2>9. Converting ints/floats to strings.</h2> 
    <p> Many times we have to convert one variable type to another. Here's one example to see why.</p>
    

Lets say we want to print the phrase "The length of the wall is 10ft." We might hope that we can do this by concatenating the variable 
<code>feet</code> with the string abbreviation "ft".  Let's see what happens when we do this.

In [47]:
 #Run this code. It causes an error.  What does the error say?
feet = 10
print("The length of the wall is " + feet + "ft.")

TypeError: can only concatenate str (not "int") to str

The issue in the above code is that we can add number types together and we can add string types together (remember this is called string concatenation) but Python doesn't understand how to add an integer with a string. So we must convert the number to a string by using the str method, as shown in the next example.

In [48]:
print ("The length of the wall is " + str(feet) + "ft.")  #str is a function that converts feet (which is an integer) to a string
                                                          #so that we can concatenate it with the surrounding text.

The length of the wall is 10ft.


---

## **Question 9!!!!!!!!!!!!!!!!!!!** 

The code below creates an error.  Fix it so that the phrase "42 seconds equals 0.7 minutes" is printed out.

In [51]:
time_in_seconds = 42
time_in_minutes = time_in_seconds / 60  #Divide seconds by 60 to convert to minutes.

print(str(time_in_seconds) + " seconds equals " + str(time_in_minutes) + " minutes")  #what should this print?   Run this code to see!

42 seconds equals 0.7 minutes


Double-click __here__ for the solution.

<!-- Joe Biden would have typed the following code:
    
print(str(time_in_seconds) + " seconds equals " + str(time_in_minutes) + " minutes")
-->

<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [SUMMARY:] The str function casts numbers to strings. Example:  str(3.5).
</div>
<hr/>

<hr/><h2 id="convert">10. Converting strings to ints/float.</h2>

<p>Sometimes, we can have a string that contains a number within it. If this is the case, we can cast that string that represents a number into an integer using <code>int()</code>:</p>

In [52]:
# Convert a string into an integer

x = int('1')
print(x)
print(type(x))
      

1
<class 'int'>


<p>But if you try to do so with a string that is not a perfect match for a number, you'll get an error. Try the following:</p>

In [53]:
# Convert a string into an integer with error

int('1 or 2 people')

ValueError: invalid literal for int() with base 10: '1 or 2 people'

<p>You can also convert strings containing floating point numbers into <i>float</i> objects:</p>

In [54]:
# Convert the string "1.2" into a float

x = float('1.2')
print(type(x))

<class 'float'>


<p>You can change the type of the object in Python; this is called typecasting. For example, you can convert an <i>integer</i> into a <i>float</i> (e.g. 2 to 2.0).</p>

In [55]:
#The variable x is an integer.  
x = 5

#Cast it to a float.
y = float(x)

#What is different about the results of these 2 print statements?
print(x)
print(y)

5
5.0


<p>When we convert an integer into a float, we don't really change the value (i.e., the significand) of the number. However, if we cast a float into an integer, we could potentially lose some information. For example, if we cast the float 1.1 to integer we will get 1 and lose the decimal information (i.e., 0.1):</p>

In [56]:
# Casting 1.1 to integer will result in loss of information

x = 1.1
y = int(x)

#Guess what is printed below.
print(x)
print(y)

1.1
1



<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [SUMMARY:] The int and float functions cast strings to numbers. Example:  int("10") and float("10.5").
</div>


---

## **Question 10!!!!!!!!!!!!!!!!!!!**  
    

The code below relies on string concatenation and so prints "3.14.8".  Run the code to make sure you believe this.  Change the code so that the sum  of x and y (7.9) is printed out instead.

In [62]:
x = "3.1"
y = "4.8"
print(float(x)+float(y))

7.9


Double-click __here__ for the solution.

<!-- 
x = "3.1"
y = "4.8"
print(float(x)+float(y))
-->

<hr/>
<h2 id="bool">11. Boolean data type</h2>

<p><i>Boolean</i> is another important type in Python. An object of type <i>Boolean</i> can take on one of two values: <code>True</code> or <code>False</code>:</p>

In [63]:
# Value true

True

True

<p>Notice that the value <code>True</code> has an uppercase "T". The same is true for <code>False</code> (i.e. you must use the uppercase "F").</p>

In [ ]:
# Value false

False

<p>When you ask Python to display the type of a boolean object it will show <code>bool</code> which stands for <i>boolean</i>:</p> 

In [ ]:
# Type of True

type(True)

In [ ]:
# Type of False

type(False)

<p>We can cast boolean objects to other data types. If we cast a boolean with a value of <code>True</code> to an integer or float we will get a one. If we cast a boolean with a value of <code>False</code> to an integer or float we will get a zero. Similarly, if we cast a 1 to a Boolean, you get a <code>True</code>. And if we cast a 0 to a Boolean we will get a <code>False</code>. Let's give it a try:</p> 

In [64]:
# Convert True to int

int(True)

1

In [65]:
# Convert 1 to boolean

bool(1)

True

In [66]:
# Convert 0 to boolean

bool(0)

False

In [67]:
# Convert True to float

float(True)

1.0

The point of the previous few examples was just to introduce you to the BOOLEAN type (True/False).

<hr>

Okay!  You're done Lab B!  Submit this!   Your file should be called LabB_PythonBasics_YourInitials.ipynb 